### PR TITLE
feat(ui): add BufWriteCmd to handle ":w" in quick menu

### LINKED ISSUE #36: Closing quick menu doesn't save unless toggle is called
Autocmd time!  Lets listen for win close events and ensure that we save upon floating window closed with ZZ/q

### PR DESCRIPTION
**Pull request by asbjornhaland**
Fix #36

@ThePrimeagen @brandoncc Can you have a look, please test this before merge.

### CODE REVIEW

File: lua/harpoon/ui.lua [hunk 1]
```diff
@@ -65,8 +64,16 @@ M.toggle_quick_menu = function()
...
+    vim.api.nvim_buf_set_name(bufh, "harpoon-menu")
     vim.api.nvim_buf_set_lines(bufh, 0, #contents, false, contents)
     vim.api.nvim_buf_set_option(bufh, "filetype", "harpoon")
+    vim.api.nvim_buf_set_option(bufh, "buftype", "acwrite")
+    vim.cmd(string.format("autocmd BufWriteCmd <buffer=%s> :lua require('harpoon.ui').on_menu_save()", bufh))
+end
+
+M.on_menu_save = function()
+    save_changes()
+    vim.cmd("set nomodified")
```
Comment (brandoncc) [inline lua/harpoon/ui.lua:76]: Interesting, so this is the piece I was missing?

File: lua/harpoon/ui.lua [hunk 2]
```diff
@@ -65,8 +64,16 @@ M.toggle_quick_menu = function()
         contents[idx] = string.format("%d %s", idx, file)
     end
 
+    vim.api.nvim_buf_set_name(bufh, "harpoon-menu")
```
Comment (brandoncc) [inline lua/harpoon/ui.lua:67]: Oh, I think this is really the one I was missing!

File: lua/harpoon/ui.lua (hunk 1 above)
Comment (asbjornhaland) [reply lua/harpoon/ui.lua:76]: If modified is not reset, you indicate the write was *not* successful, + you get error when closing nvim that you have modified buffers.

You were very close though, I think your missing piece was naming the buffer with `vim.api.nvim_buf_set_name`.

File: lua/harpoon/ui.lua (hunk 2 above)
Comment (asbjornhaland) [reply lua/harpoon/ui.lua:67]: Yes

**Comment by ThePrimeagen** [comment]
@brandoncc and @asbjornhaland have you verified this all works?  I'll merge it (no time to test currently)

Also, very interesting.  Could you write up exactly / point me to which :h commands to read on, so I understand why this works :)

**Comment by asbjornhaland** [comment]
- `:h E676` is the explanation for `buftype=acwrite` and `BufWriteCmd`
- `:h E32` is nameing the buffer

**Comment by brandoncc** [comment]
I get this when I try to reopen the menu after doing `ZZ`:

```
E5108: Error executing lua /Users/brandoncc/.vim/plugged/harpoon/lua/harpoon/ui.lua:67: Vim:E95: Buffer with this name already exists
```

Maybe we need to recycle the buffer when opening by trying to find a buffer by the same name instead of creating it? Only create if there isn't one already by the same name?

**Comment by brandoncc** [comment]
@asbjornhaland I found some issues.

`:wq` and `ZZ` work. `:w` and then `:q` doesn't, though.

Also, if you make a change and then `:q`, it tells you that you need to save your changes. I wonder if we need to set modified to false in an `on_lines` callback?

**Comment by asbjornhaland** [comment]
If you make a change, doesn't it make sense that you need to `:q!` - this is normal vim  behavior?

I will have a look at why modified is set after a `:w` - it shouldn't.

**Comment by brandoncc** [comment]
> If you make a change, doesn't it make sense that you need to `:q!` - this is normal vim behavior?
> 
> I will have a look at why modified is set after a `:w` - it shouldn't.

I might have misunderstood @ThePrimeagen, but I thought he said he didn't want to require `:q!` to close. At first I thought he was saying he _did_ want it, but then I thought he meant he didn't.

Sorry, I wasn't clear about my bug report. `:wq` saves, but `:w` and then `:q` doesn't. The old content comes back after doing `:w`,`:q`.

**Comment by ThePrimeagen** [comment]
I _do not want_ q! required

**Comment by brandoncc** [comment]
Scratch that `:w:q` bug report, it looks like it was user error.

**Pull request merged**